### ISSUE TITLE
Need a callback for "on just before position discontinuity frame"?

### ISSUE THREAD

**Issue by TheNavigateur** [google/ExoPlayer#6576]
### Searched documentation and issues
https://exoplayer.dev/playlists.html

### Question
`onPositionDiscontinuity` is dispatched apparently after the first frame of the "next video" is already rendered. I am making a "playlist" of videos whose position on screen and size can change per video. The problem with waiting for the first frame of the next video is that I see the "next video" with the wrong size and potentially wrong position for a moment before I can correct it. I cannot seem to get to an event _just before_ the first frame of the "next video" is rendered so that it will look right. (I'm using OpenGL with a surface view used by the ExoPlayer).

Is this currently possible?

Or, would I need to dig into the source code and add an event for "on just before position discontinuity frame"?

If so, can you provide some easy tips for where I would do that in the code (which functions, code branches, etc.)?

Thanks!

**Comment by tonihei** [comment]
We are currently in the process of adding such an event. Marking as an enhancement to track that publicly.

**Comment by TheNavigateur** [comment]
Thank you. However, is it easy to do myself, for the time being, as I need it urgently? I have local exoplayer source code working in my project. I will attempt to do so anyway, so any advice would be appreciated

**Comment by tonihei** [comment]
You need to send this before the playbackInfo is updated [here](https://github.com/google/ExoPlayer/blob/41b3fc110178e9a5c969a0102c5b074189065273/library/core/src/main/java/com/google/android/exoplayer2/ExoPlayerImpl.java#L730). No guarantees though about unexpected edge cases, we need to check that first.

**Comment by TheNavigateur** [comment]
Hi tonihei! Before I insert code there, are you 100% sure, as it would seem "too late", as the first frame would already be drawn by the time of that code (in the case of a position discontinuity)?

**Comment by tonihei** [comment]
This callback would not be tied to the actual rendering. Actually, none of our callbacks are because they are delivered on a different thread, that's why there can't be any guarantee of how far rendering has advanced when a callback arrives. 

The intention of this callback would just be to allow apps to know the position and state of a player before a discontinuity, so that it can be logged for analytics purposes.

If you want to have frame-accurate callbacks because you intend to do something with the frames for example, you could add a `VideoFrameMetadataListener` to the video renderer that calls a method on the rendering thread every time a frame is about to be released. 

Can you share more details about your use case to understand which of the callbacks is relevant for what you are trying to achieve?

**Comment by TheNavigateur** [comment]
Yes!

My video playlist has different video sizes and positions on screen per video.

The "callback" is for me to make the "switch layout" request _before_ the first frame of the next video is rendered, so that each video is always shown in its intended position and size. Currently, the first frame of the next video is rendering on the "current" layout, where it doesn't belong, and I cannot get any "event" to fire _before_ the first frame of the next video has already been rendered...

**Comment by TheNavigateur** [comment]
Hi @tonihei !

OK, so I have a `VideoFrameMetadataListener` with the `onVideoFrameAboutToBeRendered` method...

Do you have any idea how I would detect that this "video frame" is the "first frame of the next video"?

I can confirm that both the position discontinuity and "before player info update" happen after that first frame is already rendered (so too late). 

I am able to edit the source code locally for my project, so I can try out any suggestions you have.

**Comment by tonihei** [comment]
> Do you have any idea how I would detect that this "video frame" is the "first frame of the next video"?

That's not part of this listener unfortunately. The easiest way I can see right now is to subclass `MediaCodecVideoRenderer` and overwrite `onStreamChanged` (don't forget to call super). This method gives you the streamOffset when a new stream starts (including the first one). You can inject your own renderer version by overwriting `DefaultRenderersFactory.buildVideoRenderers` and providing this factory to the player factory.

> The "callback" is for me to make the "switch layout" request before the first frame of the next video is rendered, so that each video is always shown in its intended position and size.

Not sure it can solve your problem, but did you have a look at `AspectRatioFrameLayout` in the UI module? This layout resizes itself according to your settings and is used by default in our `PlayerView`. If you are using `PlayerView`, you can adjust the `resize_mode` parameter if needed.

**Comment by TheNavigateur** [comment]
Hi @tonihei !

Thanks, I will try this approach now.

I'm using OpenGL so the frame layout approach wouldn't work for my case.

I'll let you know how I get on.

**Comment by TheNavigateur** [comment]
Hi @tonihei !

I've been struggling, but now things seem OK so far.

`onStreamChanged` is called too early to be used directly for my use case, in fact earlier than the frame even _before_ the first frame of the "next video", which means I can't even use it directly to set e.g. a boolean for "the next `onVideoFrameAboutToBeRendered `", etc.

However, it turns out a really simple approach is working for me: `onVideoFrameAboutToBeRendered ` has a `presentationTimeUs`, which represents the first frame when it is 0. I needed to defer my "layout switch" to the next "video render" after that point in time, and it is working for me.

Other than that, I just need to iron out a few other peripheral, logical issues, and I think I'm good. I'll be testing it over the coming period.

Thanks for your help.

@TheNavigateur

**Comment by tonihei** [comment]
> onStreamChanged is called too early to be used directly for my use case,

That's correct `onStreamChanged` will be called earlier. That's why you need to store the offset time from the parameter to make use of it. 

> a presentationTimeUs, which represents the first frame when it is 0.

That isn't strictly guaranteed for all cases. If we can keep the MediaCodec because the video format is compatible to the previous one, the presentation times will just continue to increase monotonically. That's why the onStreamChanged offset that tells you at which presentation time the new video starts.

**Comment by TheNavigateur** [comment]
Hi @tonihei !

Thanks for the information. Just trying to come up with a formula for converting `offsetTime` into a "non-zero" `presentationTime` to check (alongside 0 itself) to detect whether it's the "first frame of the next video".

Can you confirm that `presentationTimeUs` for the first frame is always either 0 or incrementing at exactly the same rate as `offsetTimeUs` (i.e. that it never stays at the same non-zero value across any 2 videos)?

**Comment by tonihei** [comment]
> Can you confirm that presentationTimeUs for the first frame is always either 0 or incrementing at exactly the same rate as offsetTimeUs

OffsetTimeUs is the first presentationTimeUs of the next stream. There is no notion of incrementing timestamps involved here.

**Comment by TheNavigateur** [comment]
Hi @tonihei !

For me, `offsetTimeUs` keeps increasing indefinitely, while `presentationTimeUs` resets to 0 (every video in my case), hence my question about increment rate.

Cheers,

N

**Comment by tonihei** [comment]
Yes, indeed. It seems we already subtract the stream offset from the presentation time before calling this method (sorry for the confusion). That means listening to the callback is enough and `offsetUs` is irrelevant for your case.

**Comment by TheNavigateur** [comment]
@tonihei OK so just to clarify, I can always rely on `presentationTimeUs` to be `0` for the first frame of any video in onVideoFrameAboutToBeRendered, and only in that case, regardless, correct?

**Comment by tonihei** [comment]
That's right.
